File: CSharp/ThrustRTCLR/Reordering.cpp
#include "stdafx.h"
#include "ThrustRTCLR.h"
#include "copy.h"
#include "remove.h"
#include "unique.h"
#include "partition.h"

namespace ThrustRTCLR
{
	template<typename T>
	inline T* just_cast_it(IntPtr p)
	{
		return (T*)(void*)p;
	}

	uint32_t Native::copy_if(IntPtr p_vec_in, IntPtr p_vec_out, IntPtr p_pred)
	{
		DVVectorLike* vec_in = just_cast_it<DVVectorLike>(p_vec_in);
		DVVectorLike* vec_out = just_cast_it<DVVectorLike>(p_vec_out);
		Functor* pred = just_cast_it<Functor>(p_pred);
		return TRTC_Copy_If(*vec_in, *vec_out, *pred);
	}

	uint32_t Native::copy_if_stencil(IntPtr p_vec_in, IntPtr p_vec_stencil, IntPtr p_vec_out, IntPtr p_pred)
	{
		DVVectorLike* vec_in = just_cast_it<DVVectorLike>(p_vec_in);
		DVVectorLike* vec_stencil = just_cast_it<DVVectorLike>(p_vec_stencil);
		DVVectorLike* vec_out = just_cast_it<DVVectorLike>(p_vec_out);
		Functor* pred = just_cast_it<Functor>(p_pred);
		return TRTC_Copy_If_Stencil(*vec_in, *vec_stencil, *vec_out, *pred);
	}

	uint32_t Native::remove(IntPtr p_vec, IntPtr p_value)
	{
		DVVectorLike* vec = just_cast_it<DVVectorLike>(p_vec);
		DeviceViewable* value = just_cast_it<DeviceViewable>(p_value);
		return TRTC_Remove(*vec, *value);
	}

	uint32_t Native::remove_copy(IntPtr p_vec_in, IntPtr p_vec_out, IntPtr p_value)
	{
		DVVectorLike* vec_in = just_cast_it<DVVectorLike>(p_vec_in);
		DVVectorLike* vec_out = just_cast_it<DVVectorLike>(p_vec_out);
		DeviceViewable* value = just_cast_it<DeviceViewable>(p_value);
		return TRTC_Remove_Copy(*vec_in, *vec_out, *value);
	}

	uint32_t Native::remove_if(IntPtr p_vec, IntPtr p_pred)
	{
		DVVectorLike* vec = just_cast_it<DVVectorLike>(p_vec);
		Functor* pred = just_cast_it<Functor>(p_pred);
		return TRTC_Remove_If(*vec, *pred);
	}

	uint32_t Native::remove_copy_if(IntPtr p_vec_in, IntPtr p_vec_out, IntPtr p_pred)
	{
		DVVectorLike* vec_in = just_cast_it<DVVectorLike>(p_vec_in);
		DVVectorLike* vec_out = just_cast_it<DVVectorLike>(p_vec_out);
		Functor* pred = just_cast_it<Functor>(p_pred);
		return TRTC_Remove_Copy_If(*vec_in, *vec_out, *pred);
	}

	uint32_t Native::remove_if_stencil(IntPtr p_vec, IntPtr p_stencil, IntPtr p_pred)
	{
		DVVectorLike* vec = just_cast_it<DVVectorLike>(p_vec);
		DVVectorLike* stencil = just_cast_it<DVVectorLike>(p_stencil);
		Functor* pred = just_cast_it<Functor>(p_pred);
		return TRTC_Remove_If_Stencil(*vec, *stencil, *pred);
	}

	uint32_t Native::remove_copy_if_stencil(IntPtr p_vec_in, IntPtr p_stencil, IntPtr p_vec_out, IntPtr p_pred)
	{
		DVVectorLike* vec_in = just_cast_it<DVVectorLike>(p_vec_in);
		DVVectorLike* stencil = just_cast_it<DVVectorLike>(p_stencil);
		DVVectorLike* vec_out = just_cast_it<DVVectorLike>(p_vec_out);
		Functor* pred = just_cast_it<Functor>(p_pred);
		return TRTC_Remove_Copy_If_Stencil(*vec_in, *stencil, *vec_out, *pred);
	}

	uint32_t Native::unique(IntPtr p_vec)
	{
		DVVectorLike* vec = just_cast_it<DVVectorLike>(p_vec);
		return TRTC_Unique(*vec);
	}

	uint32_t Native::unique(IntPtr p_vec, IntPtr p_binary_pred)
	{
		DVVectorLike* vec = just_cast_it<DVVectorLike>(p_vec);
		Functor* binary_pred = just_cast_it<Functor>(p_binary_pred);
		return TRTC_Unique(*vec, *binary_pred);
	}

	uint32_t Native::unique_copy(IntPtr p_vec_in, IntPtr p_vec_out)
	{
		DVVectorLike* vec_in = just_cast_it<DVVectorLike>(p_vec_in);
		DVVectorLike* vec_out = just_cast_it<DVVectorLike>(p_vec_out);
		return TRTC_Unique_Copy(*vec_in, *vec_out);
	}

	uint32_t Native::unique_copy(IntPtr p_vec_in, IntPtr p_vec_out, IntPtr p_binary_pred)
	{
		DVVectorLike* vec_in = just_cast_it<DVVectorLike>(p_vec_in);
		DVVectorLike* vec_out = just_cast_it<DVVectorLike>(p_vec_out);
		Functor* binary_pred = just_cast_it<Functor>(p_binary_pred);
		return TRTC_Unique_Copy(*vec_in, *vec_out, *binary_pred);
	}

	uint32_t Native::unique_by_key(IntPtr p_keys, IntPtr p_values)
	{
		DVVectorLike* keys = just_cast_it<DVVectorLike>(p_keys);
		DVVectorLike* values = just_cast_it<DVVectorLike>(p_values);
		return TRTC_Unique_By_Key(*keys, *values);
	}

	uint32_t Native::unique_by_key(IntPtr p_keys, IntPtr p_values, IntPtr p_binary_pred)
	{
		DVVectorLike* keys = just_cast_it<DVVectorLike>(p_keys);
		DVVectorLike* values = just_cast_it<DVVectorLike>(p_values);
		Functor* binary_pred = just_cast_it<Functor>(p_binary_pred);
		return TRTC_Unique_By_Key(*keys, *values, *binary_pred);
	}

	uint32_t Native::unique_by_key_copy(IntPtr p_keys_in, IntPtr p_values_in, IntPtr p_keys_out, IntPtr p_values_out)
	{
		DVVectorLike* keys_in = just_cast_it<DVVectorLike>(p_keys_in);
		DVVectorLike* values_in = just_cast_it<DVVectorLike>(p_values_in);
		DVVectorLike* keys_out = just_cast_it<DVVectorLike>(p_keys_out);
		DVVectorLike* values_out = just_cast_it<DVVectorLike>(p_values_out);
		return TRTC_Unique_By_Key_Copy(*keys_in, *values_in, *keys_out, *values_out);
	}

	uint32_t Native::unique_by_key_copy(IntPtr p_keys_in, IntPtr p_values_in, IntPtr p_keys_out, IntPtr p_values_out, IntPtr p_binary_pred)
	{
		DVVectorLike* keys_in = just_cast_it<DVVectorLike>(p_keys_in);
		DVVectorLike* values_in = just_cast_it<DVVectorLike>(p_values_in);
		DVVectorLike* keys_out = just_cast_it<DVVectorLike>(p_keys_out);
		DVVectorLike* values_out = just_cast_it<DVVectorLike>(p_values_out);
		Functor* binary_pred = just_cast_it<Functor>(p_binary_pred);
		return TRTC_Unique_By_Key_Copy(*keys_in, *values_in, *keys_out, *values_out, *binary_pred);
	}

	uint32_t Native::partition(IntPtr p_vec, IntPtr p_pred)
	{
		DVVectorLike* vec = just_cast_it<DVVectorLike>(p_vec);
		Functor* pred = just_cast_it<Functor>(p_pred);
		return TRTC_Partition(*vec, *pred);
	}

	uint32_t Native::partition_stencil(IntPtr p_vec, IntPtr p_stencil, IntPtr p_pred)
	{
		DVVectorLike* vec = just_cast_it<DVVectorLike>(p_vec);
		DVVectorLike* stencil = just_cast_it<DVVectorLike>(p_stencil);
		Functor* pred = just_cast_it<Functor>(p_pred);
		return TRTC_Partition_Stencil(*vec, *stencil, *pred);
	}

	uint32_t Native::partition_copy(IntPtr p_vec_in, IntPtr p_vec_true, IntPtr p_vec_false, IntPtr p_pred)
	{
		DVVectorLike* vec_in = just_cast_it<DVVectorLike>(p_vec_in);
		DVVectorLike* vec_true = just_cast_it<DVVectorLike>(p_vec_true);
		DVVectorLike* vec_false = just_cast_it<DVVectorLike>(p_vec_false);
		Functor* pred = just_cast_it<Functor>(p_pred);
		return TRTC_Partition_Copy(*vec_in, *vec_true, *vec_false, *pred);
	}

	uint32_t Native::partition_copy_stencil(IntPtr p_vec_in, IntPtr p_stencil, IntPtr p_vec_true, IntPtr p_vec_false, IntPtr p_pred)
	{
		DVVectorLike* vec_in = just_cast_it<DVVectorLike>(p_vec_in);
		DVVectorLike* vec_stencil = just_cast_it<DVVectorLike>(p_stencil);
		DVVectorLike* vec_true = just_cast_it<DVVectorLike>(p_vec_true);
		DVVectorLike* vec_false = just_cast_it<DVVectorLike>(p_vec_false);
		Functor* pred = just_cast_it<Functor>(p_pred);
		return TRTC_Partition_Copy_Stencil(*vec_in, *vec_stencil, *vec_true, *vec_false, *pred);
	}

}

File: CSharp/ThrustRTCSharp/Reordering.cs
using ThrustRTCLR;

namespace ThrustRTCSharp
{
    public partial class TRTC
    {
        public static long Copy_If(DVVectorLike vec_in, DVVectorLike vec_out, Functor pred)
        {
            return (long)Native.copy_if(vec_in.m_cptr, vec_out.m_cptr, pred.m_cptr);
        }

        public static long Copy_If_Stencil(DVVectorLike vec_in, DVVectorLike vec_stencil, DVVectorLike vec_out, Functor pred)
        {
            return (long)Native.copy_if_stencil(vec_in.m_cptr, vec_stencil.m_cptr, vec_out.m_cptr, pred.m_cptr);
        }

        public static long Remove(DVVectorLike vec, DeviceViewable value)
        {
            return (long)Native.remove(vec.m_cptr, value.m_cptr);
        }

        public static long Remove_Copy(DVVectorLike vec_in, DVVectorLike vec_out, DeviceViewable value)
        {
            return (long)Native.remove_copy(vec_in.m_cptr, vec_out.m_cptr, value.m_cptr);
        }

        public static long Remove_If(DVVectorLike vec, Functor pred)
        {
            return (long)Native.remove_if(vec.m_cptr, pred.m_cptr);
        }

        public static long Remove_Copy_If(DVVectorLike vec_in, DVVectorLike vec_out, Functor pred)
        {
            return (long)Native.remove_copy_if(vec_in.m_cptr, vec_out.m_cptr, pred.m_cptr);
        }

        public static long Remove_If_Stencil(DVVectorLike vec, DVVectorLike stencil, Functor pred)
        {
            return (long)Native.remove_if_stencil(vec.m_cptr, stencil.m_cptr, pred.m_cptr);
        }

        public static long Remove_Copy_If_Stencil(DVVectorLike vec_in, DVVectorLike stencil, DVVectorLike vec_out, Functor pred)
        {
            return (long)Native.remove_copy_if_stencil(vec_in.m_cptr, stencil.m_cptr,vec_out.m_cptr, pred.m_cptr);
        }

        public static long Unique(DVVectorLike vec)
        {
            return (long)Native.unique(vec.m_cptr);
        }

        public static long Unique(DVVectorLike vec, Functor binary_pred)
        {
            return (long)Native.unique(vec.m_cptr, binary_pred.m_cptr);
        }

        public static long Unique_Copy(DVVectorLike vec_in, DVVectorLike vec_out)
        {
            return (long)Native.unique_copy(vec_in.m_cptr, vec_out.m_cptr);
        }

        public static long Unique_Copy(DVVectorLike vec_in, DVVectorLike vec_out, Functor binary_pred)
        {
            return (long)Native.unique_copy(vec_in.m_cptr, vec_out.m_cptr, binary_pred.m_cptr);
        }

        public static long Unique_By_Key(DVVectorLike keys, DVVectorLike values)
        {
            return (long)Native.unique_by_key(keys.m_cptr, values.m_cptr);
        }

        public static long Unique_By_Key(DVVectorLike keys, DVVectorLike values, Functor binary_pred)
        {
            return (long)Native.unique_by_key(keys.m_cptr, values.m_cptr, binary_pred.m_cptr);
        }

        public static long Unique_By_Key_Copy(DVVectorLike keys_in, DVVectorLike values_in, DVVectorLike keys_out, DVVectorLike values_out)
        {
            return (long)Native.unique_by_key_copy(keys_in.m_cptr, values_in.m_cptr, keys_out.m_cptr, values_out.m_cptr);
        }

        public static long Unique_By_Key_Copy(DVVectorLike keys_in, DVVectorLike values_in, DVVectorLike keys_out, DVVectorLike values_out, Functor binary_pred)
        {
            return (long)Native.unique_by_key_copy(keys_in.m_cptr, values_in.m_cptr, keys_out.m_cptr, values_out.m_cptr, binary_pred.m_cptr);
        }

        public static long Partition(DVVectorLike vec, Functor pred)
        {
            return (long)Native.partition(vec.m_cptr, pred.m_cptr);
        }

        public static long Partition_Stencil(DVVectorLike vec, DVVectorLike stencil, Functor pred)
        {
            return (long)Native.partition_stencil(vec.m_cptr, stencil.m_cptr, pred.m_cptr);
        }

        public static long Partition_Copy(DVVectorLike vec_in, DVVectorLike vec_true, DVVectorLike vec_false, Functor pred)
        {
            return (long)Native.partition_copy(vec_in.m_cptr, vec_true.m_cptr, vec_false.m_cptr, pred.m_cptr);
        }

        public static long Partition_Copy_Stencil(DVVectorLike vec_in, DVVectorLike stencil, DVVectorLike vec_true, DVVectorLike vec_false, Functor pred)
        {
            return (long)Native.partition_copy_stencil(vec_in.m_cptr, stencil.m_cptr, vec_true.m_cptr, vec_false.m_cptr, pred.m_cptr);
        }

    }
}
